File: core/src/adapters/export_scene_graph.cpp
#include "toporoom/adapters/export_scene_graph.hpp"

#include <algorithm>
#include <nlohmann/json.hpp>
#include <unordered_map>

namespace toporoom::adapters {
namespace {

std::array<double, 3> midpoint_meters(const domain::SceneIRPoint& start,
                                      const domain::SceneIRPoint& end) {
  return {((start.x + end.x) / 2.0) * kMmToM, 0.0,
          ((start.y + end.y) / 2.0) * kMmToM};
}

}  // namespace

SceneGraph export_scene_graph(const domain::SceneIR& scene,
                              const ports::MeshProjection& /*meshes*/) {
  SceneGraph graph;
  for (const auto& storey : scene.storeys) {
    SceneGraphNode storey_node;
    storey_node.name = "Storey_" + storey.id;
    storey_node.toporoom_id = storey.id;
    storey_node.kind = "storey";
    storey_node.translation_meters = {0.0, storey.elevation_mm * kMmToM, 0.0};

    std::vector<SceneGraphNode> extra;
    for (const auto& wall : storey.walls) {
      const std::string wall_name = "Wall_" + wall.id;
      storey_node.children.push_back(wall_name);
      SceneGraphNode wall_node;
      wall_node.name = wall_name;
      wall_node.toporoom_id = wall.id;
      wall_node.kind = "wall";
      wall_node.translation_meters = midpoint_meters(wall.start, wall.end);
      for (const auto& opening : wall.openings) {
        const std::string opening_name = "Opening_" + opening.id;
        wall_node.children.push_back(opening_name);
        SceneGraphNode opening_node;
        opening_node.name = opening_name;
        opening_node.toporoom_id = opening.id;
        opening_node.kind = "opening";
        extra.push_back(std::move(opening_node));
      }
      extra.insert(extra.begin(), std::move(wall_node));
    }
    for (const auto& room : storey.rooms) {
      const std::string room_name = "Room_" + room.id;
      storey_node.children.push_back(room_name);
      SceneGraphNode room_node;
      room_node.name = room_name;
      room_node.toporoom_id = room.id;
      room_node.kind = "room";
      extra.push_back(std::move(room_node));
    }
    graph.nodes.push_back(std::move(storey_node));
    graph.nodes.insert(graph.nodes.end(), extra.begin(), extra.end());
  }
  return graph;
}

std::string scene_graph_to_json(const SceneGraph& graph) {
  nlohmann::json nodes = nlohmann::json::array();
  for (const auto& node : graph.nodes) {
    nlohmann::json j = {{"name", node.name},
                        {"extras",
                         {{"toporoomId", node.toporoom_id}, {"kind", node.kind}}},
                        {"children", node.children}};
    if (node.translation_meters) {
      j["translationMeters"] = *node.translation_meters;
    }
    nodes.push_back(std::move(j));
  }
  nlohmann::json root = {{"generator", graph.generator},
                         {"sourceUnits", graph.source_units},
                         {"units", graph.units},
                         {"nodes", nodes}};
  return root.dump();
}

std::string export_gltf_json(const domain::SceneIR& scene,
                             const ports::MeshProjection& meshes) {
  const SceneGraph graph = export_scene_graph(scene, meshes);
  nlohmann::json nodes = nlohmann::json::array();
  std::unordered_map<std::string, int> index_by_name;
  for (int i = 0; i < static_cast<int>(graph.nodes.size()); ++i) {
    index_by_name[graph.nodes[static_cast<std::size_t>(i)].name] = i;
  }

  nlohmann::json root_nodes = nlohmann::json::array();
  for (int i = 0; i < static_cast<int>(graph.nodes.size()); ++i) {
    const auto& node = graph.nodes[static_cast<std::size_t>(i)];
    nlohmann::json gltf_node = {
        {"name", node.name},
        {"extras", {{"toporoomId", node.toporoom_id}, {"kind", node.kind}}}};
    if (!node.children.empty()) {
      nlohmann::json children = nlohmann::json::array();
      for (const auto& child : node.children) {
        children.push_back(index_by_name.at(child));
      }
      gltf_node["children"] = children;
    }
    if (node.translation_meters) {
      gltf_node["translation"] = *node.translation_meters;
    }
    for (const auto& solid : meshes.solids) {
      if (solid.node_hint == node.name && !solid.vertices_mm.empty()) {
        gltf_node["mesh"] = 0;
        break;
      }
    }
    if (node.kind == "storey") {
      root_nodes.push_back(i);
    }
    nodes.push_back(std::move(gltf_node));
  }

  nlohmann::json gltf = {{"asset",
                          {{"version", "2.0"}, {"generator", kCompilerGenerator}}},
                         {"scene", 0},
                         {"scenes", nlohmann::json::array({{{"nodes", root_nodes}}})},
                         {"nodes", nodes}};

  std::vector<double> positions;
  for (const auto& solid : meshes.solids) {
    for (double value : solid.vertices_mm) {
      positions.push_back(value * kMmToM);
    }
  }
  if (!positions.empty()) {
    std::array<double, 3> maxv{-1e300, -1e300, -1e300};
    std::array<double, 3> minv{1e300, 1e300, 1e300};
    for (std::size_t i = 0; i + 2 < positions.size(); i += 3) {
      for (int c = 0; c < 3; ++c) {
        maxv[static_cast<std::size_t>(c)] =
            std::max(maxv[static_cast<std::size_t>(c)], positions[i + static_cast<std::size_t>(c)]);
        minv[static_cast<std::size_t>(c)] =
            std::min(minv[static_cast<std::size_t>(c)], positions[i + static_cast<std::size_t>(c)]);
      }
    }
    gltf["meshes"] = nlohmann::json::array(
        {{{"primitives", nlohmann::json::array({{{"attributes", {{"POSITION", 0}}}}})}}});
    gltf["accessors"] = nlohmann::json::array({{{"bufferView", 0},
                                                {"componentType", 5126},
                                                {"count", static_cast<int>(positions.size() / 3)},
                                                {"type", "VEC3"},
                                                {"max", maxv},
                                                {"min", minv}}});
  }
  return gltf.dump();
}

}  // namespace toporoom::adapters

File: packages/adapter-export-gltf/src/export-scene-graph.ts
import type { SceneIR } from "@toporoom/domain-floorplan";
import type { MeshProjection } from "@toporoom/ports";

export const COMPILER_GENERATOR = "toporoom-compiler/0.1.0";
export const MM_TO_M = 0.001;

export interface SceneGraphExtras {
  toporoomId: string;
  kind: "storey" | "wall" | "room" | "opening";
}

export interface SceneGraphNode {
  name: string;
  extras: SceneGraphExtras;
  translationMeters?: [number, number, number];
  children: string[];
}

export interface SceneGraph {
  generator: string;
  sourceUnits: "mm";
  units: "m";
  nodes: SceneGraphNode[];
}

export interface GltfJson {
  asset: { version: "2.0"; generator: string };
  scene: number;
  scenes: { nodes: number[] }[];
  nodes: {
    name: string;
    extras: SceneGraphExtras;
    children?: number[];
    mesh?: number;
    translation?: [number, number, number];
  }[];
  meshes?: {
    primitives: { attributes: { POSITION: number }; indices?: number }[];
  }[];
  accessors?: {
    bufferView: number;
    componentType: number;
    count: number;
    type: "VEC3" | "SCALAR";
    max?: number[];
    min?: number[];
  }[];
  bufferViews?: { buffer: number; byteOffset: number; byteLength: number }[];
  buffers?: { byteLength: number; uri: string }[];
}

export function exportSceneGraph(
  scene: SceneIR,
  meshes: MeshProjection,
): SceneGraph {
  const nodes: SceneGraphNode[] = [];
  for (const storey of scene.storeys) {
    const storeyName = `Storey_${storey.id}`;
    const children: string[] = [];
    for (const wall of storey.walls) {
      const wallName = `Wall_${wall.id}`;
      children.push(wallName);
      nodes.push({
        name: wallName,
        extras: { toporoomId: wall.id, kind: "wall" },
        children: wall.openings.map((opening) => `Opening_${opening.id}`),
        translationMeters: midpointMeters(wall.start, wall.end),
      });
      for (const opening of wall.openings) {
        nodes.push({
          name: `Opening_${opening.id}`,
          extras: { toporoomId: opening.id, kind: "opening" },
          children: [],
        });
      }
    }
    for (const room of storey.rooms) {
      const roomName = `Room_${room.id}`;
      children.push(roomName);
      nodes.push({
        name: roomName,
        extras: { toporoomId: room.id, kind: "room" },
        children: [],
      });
    }
    nodes.unshift({
      name: storeyName,
      extras: { toporoomId: storey.id, kind: "storey" },
      translationMeters: [0, storey.elevationMm * MM_TO_M, 0],
      children,
    });
  }
  void meshes;
  return {
    generator: COMPILER_GENERATOR,
    sourceUnits: "mm",
    units: "m",
    nodes,
  };
}

export function exportGltfJson(
  scene: SceneIR,
  meshes: MeshProjection,
): GltfJson {
  const graph = exportSceneGraph(scene, meshes);
  const indexByName = new Map(graph.nodes.map((node, index) => [node.name, index]));
  const nodes = graph.nodes.map((node) => {
    const gltfNode: GltfJson["nodes"][number] = {
      name: node.name,
      extras: node.extras,
    };
    if (node.children.length > 0) {
      gltfNode.children = node.children.map((child) => {
        const index = indexByName.get(child);
        if (index === undefined) {
          throw new Error(`Missing child node ${child}`);
        }
        return index;
      });
    }
    if (node.translationMeters) {
      gltfNode.translation = node.translationMeters;
    }
    const solid = meshes.solids.find(
      (candidate) => candidate.nodeHint === node.name && candidate.verticesMm.length > 0,
    );
    if (solid) {
      gltfNode.mesh = 0;
    }
    return gltfNode;
  });

  const positions: number[] = [];
  for (const solid of meshes.solids) {
    for (const value of solid.verticesMm) {
      positions.push(value * MM_TO_M);
    }
  }
  const bytes = float32Buffer(positions);
  const max = [-Infinity, -Infinity, -Infinity];
  const min = [Infinity, Infinity, Infinity];
  for (let i = 0; i < positions.length; i += 3) {
    max[0] = Math.max(max[0]!, positions[i]!);
    max[1] = Math.max(max[1]!, positions[i + 1]!);
    max[2] = Math.max(max[2]!, positions[i + 2]!);
    min[0] = Math.min(min[0]!, positions[i]!);
    min[1] = Math.min(min[1]!, positions[i + 1]!);
    min[2] = Math.min(min[2]!, positions[i + 2]!);
  }
  const vertexCount = positions.length / 3;

  const rootNodes = graph.nodes
    .map((node, index) => ({ node, index }))
    .filter(({ node }) => node.extras.kind === "storey")
    .map(({ index }) => index);

  return {
    asset: { version: "2.0", generator: COMPILER_GENERATOR },
    scene: 0,
    scenes: [{ nodes: rootNodes }],
    nodes,
    ...(vertexCount > 0
      ? {
          meshes: [
            {
              primitives: [{ attributes: { POSITION: 0 } }],
            },
          ],
          accessors: [
            {
              bufferView: 0,
              componentType: 5126,
              count: vertexCount,
              type: "VEC3",
              max,
              min,
            },
          ],
          bufferViews: [
            { buffer: 0, byteOffset: 0, byteLength: bytes.byteLength },
          ],
          buffers: [
            {
              byteLength: bytes.byteLength,
              uri: `data:application/octet-stream;base64,${bytesToBase64(bytes)}`,
            },
          ],
        }
      : {}),
  };
}

function midpointMeters(
  start: { x: number; y: number },
  end: { x: number; y: number },
): [number, number, number] {
  return [
    ((start.x + end.x) / 2) * MM_TO_M,
    0,
    ((start.y + end.y) / 2) * MM_TO_M,
  ];
}

function float32Buffer(values: number[]): Uint8Array {
  const buffer = new ArrayBuffer(values.length * 4);
  new Float32Array(buffer).set(values);
  return new Uint8Array(buffer);
}

function bytesToBase64(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("base64");
}
